decentraland/unity-renderer
# fix: some combinations of properties for primitives, GLTF, NFT and  OnPointerEvents for ecs7

## Context

While we were developing the new ECS7, we develop the components quickly to be able to iterate quickly and have a first version where we could start coding some ECS7 sample scenes. Those components weren't tested roughly and we created the components of the shape without the OnPointerDown component.

Thes shapes were working fine with the OnPointerDown component but there were certain combinations of values in the properties of the model that weren't working as expected.

Also, the OnPointerDown event wasn't showing the key assigned as the ECS6.

Since we already have the ECS7 in an internal testing phase, all the components needed to be changed at once to maintain compatibility with the current version

In this PR we are tracking the issues with the combination of certain values ( In primitives, GLTF, and NFT, the key of the OnPointer event not showing and improve the performance reducing the number of colliders needed when multiple OnPointer events are in the same entity.



## What does this PR change?

- Fixes all possible combinations of the shape models. Now `withCollisions`, `isPointerBlocker ` and `visible `work independ. This means that you can combine them as you like and they will work always. Prior to this PR not all combinations were working
- Fixes an issue in ECS6 and ECS7 where primitives with the attributes `withCollisions = true` and `isPointerBlocker = false` where not renderer 
- Fixes the key not being shown when you hover on an entity that has an OnPointerEvent 
- Primitives now have their own rendering Game Object instead of the components (mesh renderer, mesh filter) being added to the entity to match parity with GLTF and NFTs
- NFT prefabs have the box collider adjusted to their mesh ( Only the ECS7 prefabs, NFTs have different prefabs for ECS6 and ECS7 since in ECS7 the error part is instantiated only when they fail, ECS6 instantiated it at the beginning and make them invisible. Performance wise is better to instantiate it only when they fail)
- Now all the OnPointerEvents ( like OnPointerDown and OnPointerUp) share the same collider instead of creating different colliders  
- Implement correct disposing of the Cancellation tokens for NFTS



## What does this PR not pretend?

Implement integration test, since that will be done in another PR. Making an integration test would require lots of changes in other systems and will make this PR more unreadable and bigger

## How to test the changes?

Can't be tested easily

1.- Download the scene https://github.com/decentraland/ecs7-template  (recommended branch:  feat/cube-spawner)
2.- Follow the instructions and start the scene with npm start
3.- Use this URL: http://127.0.0.1:8000/?position=0%2C0&ENABLE_ECS7=&renderer-branch=fix%2Fbox-shape-ecs7&kernel-branch=main&SCENE_DEBUG_PANEL=&realm=v1%7E127.0.0.1%3A8000
4.- Change the code to see if everything works as expected

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
